### ISSUE TITLE
Implement More Widescreen Hacks

### ISSUE THREAD

**Issue by cheatfreak47** [FIX94/Nintendont#272]
Something I've noticed is that there is a metric ton of available widescreen hacks on the Dolphin wiki that happen to just work with various games. I'm planning on going through and testing each of these myself, for the record and will update the issue with more information as I test each one.

The idea is that I might find patches that work incredibly well for specific games either better than the existing ones or ones that take a previously not working game, and make it work. If so, those patches should be quite suitable to be implemented into Nintendont itself.

I'll also add more games with working patches as I find them. **You can find any of the codes/patches I'm referring to on the Dolphin Wiki pages for said games.**

Otherwise this issue is just a generic "more widescreen patches" enhancement request, but I will try to maintain this as I actively search out and test more patches

### Legend:
None = untested
**Bold** = Works and Tested
**Bold + Italics** = Is definitely good enough to be added to Nintendont
~~Strikethrough~~ = Doesn't seem to work 
^ = note

### List of ones checked:
_**Kirby Air Ride**_ ^ v1 and v2 do not seem to be much different, but I prefer v2 myself it just looks better somehow I think
_**Shadow the Hedgehog**_ ^Minor visual distortion during chaos control, works great otherwise
_**Wario World**_ ^works beautifully
_**Paper Mario The Thousand Year Door**_ ^definitely fixes a lot of clipping related problems
~~Sonic Adventure DX~~ ^crashes once you start playing I guess
~~The Legend of Zelda: Four Swords Adventure~~ ^crashes to exception error?!
The Legend of Zelda Twilight Princess
The Legend of Zelda Wind Waker 
Tony Hawk's Pro Skater 3 
Bomberman Generation
Dragon Ball Z Sagas
Billy Hatcher and the Giant Egg
Fire Emblem: Path of Radiance
Bomberman Jetters
Super Mario Sunshine

### Implemented Already:
N64 Emulators
Mario Kart Double Dash
Animal Crossing and e+
Super Smash Bros. Melee
1080 Avalanche 
Pikmin
Pikmin 2
Metroid Prime
Metroid Prime 2
Need for Speed Hot Pursuit 2
Luigi's Mansion
Sonic Gems Collection [Sonic R]
NFL Blitz 2002, 2003, Pro
Skies of Arcadia Legends
Sonic Heroes

**Comment by carnage702** [comment]
most widescreens hacks might not work straight away due to nintendont changing some adresses and such and need to be modified.

Also make sure nintendont default widescreen hack doesnt work already i remenber wario world default widescreen from nintendont working great on it so there is no point of forcing a custom widescreen hack on it.

**Comment by cheatfreak47** [comment]
Will do, I plan on looking into more of these tomorrow, so far Kirby Air Ride, Shadow, and Paper Mario are definitely improvments over the standard widescreen hacks

Kirby and Shadow does nothing with the existing hack
Paper Mario seems to have horrendous clipping issues

**Comment by carnage702** [comment]
preety sure sunshine also works great with nintendont default widescreen hack among others i would say first try the default nintendont widescreen hack then dolpgins if nothing changes then there is no point on porting the code.

**Comment by cheatfreak47** [comment]
@carnage702 Yeah I know, but that list is just of games I found had patches, only bolded and italicised ones are ones I'm recommending actually be added. Ones without any formatting I haven't actually manually checked yet. 

Tomorrow when I have time, I'll be checking the following things for each game and patch before posting a result:

-how the game's existing Widescreen hack behaves
-how the patch behaves compared to how the widescreen hack behaves
-play for an extended period with the patch spread across as much of a game as I can feasibly do, checking for potential side effects

Last I checked wario world had clipping issues but maybe it's since been fixed (haven't played it in a while)
As for Sunshine has indeed improved greatly with the built in hacks but still has some issues with the mirrors in gelato beach, I'd like to test this widescreen hack to see if this corrects it before I write it off

**Comment by cheatfreak47** [comment]
Alright so I went ahead and checked Wario World and simply put, you are wrong. @carnage702  
It has... frankly awful clipping on the edges of the screen. (even worse than some other games)

The code corrects this rather gracefully, and should be used instead.

**Comment by carnage702** [comment]
i wasnt wrong i played warioware long time ago and the widescreen patch was changed several times to be more compatible with more games i played it when the game was compatible wich was when they added the external font support so like 200+ revs ago lol.

**Comment by cheatfreak47** [comment]
@carnage702 The current Widescreen hack with the latest release version of Nintendont causes **Wario World** to have horrible clipping and pop-in on the edges of the screen. I literally just checked myself not even 20 minutes ago. (Wario World, not to be confused with WarioWare Inc. Mega Party Game$, which iirc doesn't have any 3D elements, but I haven't tested the Widescreen hack with it, so I'll look into it eventually, might be something there)

**Comment by carnage702** [comment]
no i meant wario world the 3d platformer like i said i tested it long long ago and the widescreen patch was changed so manny times i dont doubt it changed across several games.

**Comment by cheatfreak47** [comment]
@carnage702 Well either you're mistaken or something changed since then that broke it. 
This is what the current widescreen hack does on Wario World. Yes, it's widescreen, but the clipping is so bad, it takes out you of the game.
https://u.teknik.io/MOUtn.mp4

As you can see, pretty nasty clipping problems.

**Comment by carnage702** [comment]
im not mistaken  some games even lost the widescreen hack altogether while others gained the hack itself was changed numerous times for reaching more games while breaking others so yeah it wasnt like that whn i tryed but like i said it totaly changed alot the widescreen ahck to reach more games.

**Comment by cheatfreak47** [comment]
I see, I guess some commit along the way may have just broken that one then, from the sound of things. Either way, the code I linked works pefectly for Wario World, fixes all the clipping problems entirely, which is why I'm saying it'd be nice if it could be implemented as an automatic patch in [PatchWidescreen.c](https://github.com/FIX94/Nintendont/blob/master/kernel/PatchWidescreen.c)

**Comment by GerbilSoft** [comment]
@cheatfreak47 I attempted to implement the Shadow code in PatchWidescreen.c, but it simply didn't do anything. That's why I didn't include it in the exFAT branch. (The Sonic Heroes code worked perfectly [in-level only], and I was able to port it to the JP version, but it did nothing when I tried porting it to the PAL version, so I probably missed something.)

I'll check the other games you listed, but I haven't played many of them so it'll take some time to test those codes.

Both THPS4 and Jimmy Neutron weren't listed on https://en.wikipedia.org/wiki/List_of_GameCube_games_with_alternate_display_modes , which is why I didn't add those to the exception list. I'll add those in when I test the other widescreen codes.

**Comment by cheatfreak47** [comment]
@GerbilSoft Strange, any idea why the code would work as a GCT but not as a hardcoded patch? Seems like there'd be less limitations on the latter, but what do I know? `¯\(º_o)/¯`

**Comment by GerbilSoft** [comment]
@cheatfreak47 If the code modifies a value in .data or allocated on the heap, then a hard-coded version won't work, since the hard-coded one applies the change on .dol load only. That value will then get overwritten later, so the patch has no effect.

You'd need to dig deeper into the code to figure out where it sets that value, then modify *that* code to set the right value.

**Comment by cheatfreak47** [comment]
Pardon me if this is a bit of a misguided suggestion but... 
has an internally stored GCT for these widescreen patches been considered? IE, a GCT code that's just packed in nintendont that's loaded in like a standard code to load in when the appropriate title has Widescreen Hack enabled and/or the code is appended to whatever GCT the user has (if they use one for actual cheats)

Just a thought, would probably save effort in trying to convert codes from Gecko Code format to Hard-coded patches if something like that could be implemented.

**Comment by GerbilSoft** [comment]
I was thinking of a custom "widescreen hack" patch format that could be used to move all of these widescreen hacks out of the main code so it'd be easier to add more hacks without having to implement it manually, but that would basically be the same as GCTs.

I suppose that's an option, but for now, the manual patching is preferred because of the way certain games handle it (e.g. N64 emulation).

**Comment by ArgonUK** [comment]
Hi everyone,

Enable 16x9 function in Nintendont works very rarely.

Is there any easy way to enable widescreen hacks found at 
https://forums.dolphin-emu.org/Thread-unofficial-widescreen-development-topic-archive

Most gamecube games have a 16x9 support enabled hacks, which only work on Dolphin emulator. Is there a way to write a calculator, which would add necessary offsets to codes to make them compatible with Nintendont? 

I would like to request implementation of the widescreen hacks for the following games:

**Metal Gear: Twin Snakes
Star Wars Rogue Leader: Rogue Squadron II
Star Wars Rogue Leader: Rogue Squadron III
Baten Kaitos: Eternal Wings and the Lost Ocean
Baten Kaitos Origins
Chibi-Robo!
Harvest Moon A Wonderful Life 
Kirby Air Ride
Luigi's Mansion
P.N.03
Paper Mario: The Thousand-Year Door
Smuggler's Run: Warzones
Star Fox: Assault
Super Mario Strikers
Wario World**

The hack have been produced for dolphin, now they need to be implemented into Nintendont...

**Comment by SevenWillis** [comment]
Mega Man Network Transmission also has the clipping issue. :/
